pkg/controller/remotemachineset/remotemachineset_controller.go

```diff
@@ -0,0 +1,508 @@
+	if err != nil {
+		c.logger.Errorf("error retrieving clusterDeployment from cluster: %v", err)
+	} else {
+		c.enqueueClusterDeployment(clusterDeployment)
+	}
+}
+
+// Run runs c; will not return until stopCh is closed. workers determines how
+// many machine sets will be handled in parallel.
+func (c *Controller) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	c.logger.Infof("Starting remotemachineset controller")
+	defer c.logger.Infof("Shutting down remotemachineset controller")
+
+	if !controller.WaitForCacheSync("remotemachineset", stopCh, c.clusterDeploymentSynced) {
+		c.logger.Errorf("could not sync caches for remotemachineset controller")
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (c *Controller) enqueue(clusterDeployment *cov1.ClusterDeployment) {
+	key, err := controller.KeyFunc(clusterDeployment)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", clusterDeployment, err))
+		return
+	}
+	c.queue.Add(key)
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (c *Controller) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.syncHandler(key.(string))
+	c.handleErr(err, key.(string))
+
+	return true
+}
+
+func (c *Controller) handleErr(err error, key string) {
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	logger := c.logger.WithField("clusterdeployment", key)
+
+	if c.queue.NumRequeues(key) < maxRetries {
+		logger.Infof("error syncing cluster deployment: %v", err)
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	logger.Infof("dropping cluster deployment out of the queue: %v", err)
+	c.queue.Forget(key)
+}
+
+func isMasterMachineSet(machineSet *cov1.ClusterMachineSet) bool {
+	return machineSet.NodeType == cov1.NodeTypeMaster
+}
+
+func (c *Controller) buildRemoteClusterClient(clusterDeployment *cov1.ClusterDeployment) (clusterapiclient.Interface, error) {
+	// Load the kubeconfig secret for communicating with the remote cluster:
+
+	// Step 1: Get the secret ref from the cluster
+	//
+	// TODO: The cluster controller should watch secrets until it
+	// sees a kubeconfig secret named after the cluster and then
+	// update cluster status with a reference to that
+	// secret. Until then we'll look for a
+	// "clustername-kubeconfig" secret.
+	//
+	// clusterConfig := cluster.Status.AdminKubeconfig
+	// if clusterConfig == nil {
+	// 	return nil, fmt.Errorf("unable to retrieve AdminKubeConfig from cluster status")
+	// }
+	secretName := clusterDeployment.Name + "-kubeconfig"
```

Comment:
@joelddiaz so I tested this manually and realized that we are now using clusterDeployment.Spec.ClusterID instead of clusterDeployment.Name, please update.

Reply:
rebased and modified to use <spec.clusterid>-kubeconfig as the secret name

Reply:
This is fine for now, but long term I think it would be better to use cluster name in the secret name rather than cluster ID.